shotover/src/hot_reload_client.rs

```diff
@@ -0,0 +1,218 @@
+                format!(
+                    "Failed to connect to hot reload server at: {}",
+                    self.socket_path
+                )
+            })?;
+
+        let (reader, mut writer) = stream.into_split();
+        let mut reader = BufReader::new(reader);
+
+        // Send request
+        let request_json =
+            serde_json::to_string(&request).context("Failed to serialize request")?;
+
+        debug!("Sending request: {}", request_json);
+        writer
+            .write_all(request_json.as_bytes())
+            .await
+            .context("Failed to send request")?;
+
+        // Read response
+        let response: Response = read_json(&mut reader).await?;
+        debug!("Received response: {:?}", response);
+
+        Ok(response)
+    }
+}
+
+// Reuse the same JSON parsing logic as server
+async fn read_json<T: DeserializeOwned, R: AsyncReadExt + Unpin>(reader: &mut R) -> Result<T> {
+    let mut received_bytes = Vec::new();
+    loop {
+        let mut buf = [0u8; 1024];
+        let n = reader
+            .read(&mut buf)
+            .await
+            .context("Failed to read from stream")?;
+        if n == 0 {
+            return Err(anyhow::anyhow!(
+                "Connection closed before full JSON message received"
+            ));
+        }
+        received_bytes.extend_from_slice(&buf[..n]);
+        match serde_json::from_slice(&received_bytes) {
+            Ok(response) => return Ok(response),
+            Err(e) if e.is_eof() => continue,
+            Err(e) => return Err(e).context("Failed to parse JSON"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn test_client_timeout() {
+        // Use a path that will cause the connection to hang, not fail immediately
+        // This simulates a server that accepts connections but doesn't respond
+        let client =
+            UnixSocketClient::new("/dev/null".to_string()).with_timeout(Duration::from_millis(100));
+
+        let result = client.send_request(Request::SendListeningSockets).await;
+        assert!(result.is_err());
+        let error_msg = result.unwrap_err().to_string();
+
+        // The error could be either a timeout or a connection error, both are valid
+        assert!(
+            error_msg.contains("timed out") || error_msg.contains("Failed to connect"),
+            "Expected timeout or connection error, got: {}",
+            error_msg
+        );
+    }
+
+    #[tokio::test]
+    async fn test_client_connection_error() {
+        let client = UnixSocketClient::new("/nonexistent/path.sock".to_string());
+
+        let result = client.send_request(Request::SendListeningSockets).await;
+        assert!(result.is_err());
+        assert!(
+            result
+                .unwrap_err()
+                .to_string()
+                .contains("Failed to connect")
+        );
+    }
+
+    #[tokio::test]
+    async fn test_client_with_timeout_builder() {
```

Comment:
This is testing implementation details, `timeout_duration` is a private field, so any code outside of this module cannot access it.

So I'd prefer to remove this test.